lightning/restore/restore_test.go

```diff
@@ -148,6 +157,88 @@ func (s *restoreSuite) TestDoChecksum(c *C) {
 	})
 }
 
+func (s *restoreSuite) TestDoChecksumParallel(c *C) {
+	db, mock, err := sqlmock.New()
+	c.Assert(err, IsNil)
+	defer db.Close()
+
+	mock.ExpectQuery("\\QSELECT VARIABLE_VALUE FROM mysql.tidb WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+		WillReturnRows(sqlmock.NewRows([]string{"VARIABLE_VALUE"}).AddRow("10m"))
+	mock.ExpectExec("\\QUPDATE mysql.tidb SET VARIABLE_VALUE = ? WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+		WithArgs("100h0m0s").
+		WillReturnResult(sqlmock.NewResult(1, 1))
+	for i := 0; i < 5; i++ {
+		mock.ExpectQuery("\\QADMIN CHECKSUM TABLE `test`.`t`\\E").
+			WillDelayFor(100 * time.Millisecond).
+			WillReturnRows(
+				sqlmock.NewRows([]string{"Db_name", "Table_name", "Checksum_crc64_xor", "Total_kvs", "Total_bytes"}).
+					AddRow("test", "t", 8520875019404689597, 7296873, 357601387),
+			)
+	}
+	mock.ExpectExec("\\QUPDATE mysql.tidb SET VARIABLE_VALUE = ? WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+		WithArgs("10m").
+		WillReturnResult(sqlmock.NewResult(2, 1))
+
+	ctx := MockDoChecksumCtx()
+
+	var wg sync.WaitGroup
+	wg.Add(5)
+	for i := 0; i < 5; i++ {
+		go func() {
+			checksum, err := DoChecksum(ctx, db, "`test`.`t`")
+			c.Assert(err, IsNil)
+			c.Assert(*checksum, DeepEquals, RemoteChecksum{
+				Schema:     "test",
+				Table:      "t",
+				Checksum:   8520875019404689597,
+				TotalKVs:   7296873,
+				TotalBytes: 357601387,
+			})
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	c.Assert(mock.ExpectationsWereMet(), IsNil)
+	mock.ExpectClose()
+}
+
+func (s *restoreSuite) TestIncreaseGCLifeTimeFail(c *C) {
+	db, mock, err := sqlmock.New()
+	c.Assert(err, IsNil)
+	defer db.Close()
+
+	for i := 0; i < 5; i++ {
+		mock.ExpectQuery("\\QSELECT VARIABLE_VALUE FROM mysql.tidb WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+			WillReturnRows(sqlmock.NewRows([]string{"VARIABLE_VALUE"}).AddRow("10m"))
+		mock.ExpectExec("\\QUPDATE mysql.tidb SET VARIABLE_VALUE = ? WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+			WithArgs("100h0m0s").
+			WillReturnError(errors.Annotate(context.Canceled, "update gc error"))
+	}
+	// This recover GC Life Time SQL should not be executed
+	mock.ExpectExec("\\QUPDATE mysql.tidb SET VARIABLE_VALUE = ? WHERE VARIABLE_NAME = 'tikv_gc_life_time'\\E").
+		WithArgs("10m").
+		WillReturnResult(sqlmock.NewResult(1, 1))
+
+	ctx := MockDoChecksumCtx()
+	var wg sync.WaitGroup
+	wg.Add(5)
+	for i := 0; i < 5; i++ {
+		go func() {
+			_, err = DoChecksum(ctx, db, "`test`.`t`")
+			c.Assert(err, ErrorMatches, "update GC lifetime failed: update gc error: context canceled")
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	// validate no more redundant GC recover
+	err = mock.ExpectationsWereMet()
+	c.Assert(err, ErrorMatches, "there is a remaining expectation.*\n.*\n.*\n.*\n.*\n.*\n.*")
```

Comment:
😕 How about just don't add those remaining expectations?

Reply:
I tested that with this mock database, `db.Exec("some unexpected SQL")` will not fail. I guess if I didn't specify there's a remaining expectations, it won't be detected.

My test is removing `ExpectExec` of UPDATE and adding `db.Exec` of this update at end, and this test passes.

Reply:
@lance6716 That's odd, because if there's remaining expectations [it _will_ fail](https://github.com/DATA-DOG/go-sqlmock/blob/c05e0c0b9787ee10ae6f1ee6f5a5a70323bd6fcc/sqlmock.go#L174). 

Perhaps add an `ExpectClose()` and close the database instead?

Reply:
@kennytm Unexpected exec fails on the exec itself, with no effect on following `mock.ExpectationsWereMet()`'s success. If `DoChecksum` restores GC Life Time as an unexpected exec and doesn't reveal this error(perhaps this restore error is overwritten by update error), unit test can't detect that.

Of course by peeking into `DoChecksum` source code, we know unexpected exec occures in `defer` thus won't be overwritten, but I think unit test should not rely on source code check.